compiler/rustc_typeck/src/check/method/suggest.rs

```diff
@@ -346,19 +346,26 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                     );
                 }
 
-                if let Some(def) = actual.ty_adt_def() {
-                    if let Some(full_sp) = tcx.hir().span_if_local(def.did()) {
-                        let def_sp = tcx.sess.source_map().guess_head_span(full_sp);
-                        err.span_label(
-                            def_sp,
-                            format!(
-                                "{} `{}` not found {}",
-                                item_kind,
-                                item_name,
-                                if def.is_enum() && !is_method { "here" } else { "for this" }
-                            ),
+                let span_and_label_suffix = match actual.kind() {
+                    ty::Param(param_type) => {
+                        let generics = self.tcx.generics_of(
+                            tcx.hir()
+                                .body_owner_def_id(hir::BodyId { hir_id: self.body_id })
+                                .to_def_id(),
                         );
+                        let type_param = generics.type_param(param_type, self.tcx);
+                        Some((self.tcx.def_span(type_param.def_id), "for this"))
```

Comment:
We have a def-id, do you think it's useful to use `self.def_kind(def_id).descr()` so we can say "for this type parameter" or "for this enum"  below?

Reply:
I like the former suggestion "for this type parameter". I will implement it.

Reply:
Hmm. I am beginning to think it might be redundant.

Reply:
Hm, why so? I don't know why this is more clear (current):
```
LL | struct Foo {
   |        --- method `clone` not found for this
```

Than this (with the change):
```
LL | struct Foo {
   |        --- method `clone` not found for this struct
```

The "this" feels incomplete.

Reply:
I think at least we should be saying "for this type parameter" in this case, even if we don't add the `def_kind(def_id).descr()` in general.

Reply:
Ok. I will implement it 👍